doc/mgmt/Management Framework.md

```diff
@@ -0,0 +1,2029 @@
+
+REST server will integrate with Linux PAM to authenticate and authorize the user. PAM may internally use native user database or TACACS+ server based on system configuration. REST write requests will be allowed only if the user belong to admin group. Only read operations will be allowed for other users.
+
+Performing TACACS+ authentication for every REST request can slow down the APIs. This will be optimized through JSON Web Token (JWT) or a similar mechanism in future release.
+
+###### 3.2.2.4.13 Error Response
+
+REST server sends back HTTP client error (4xx) or server error (5xx) status when request processing
+fails. Response status and payload will be as per RESTCONF specifications - [RCF8040, section7](https://tools.ietf.org/html/rfc8040#page-73).
+Error response data will be a JSON with below structure. Response Content-Type will be
+"application/yang-data+json".
+
+    +---- errors
+         +---- error*
+              +---- error-type       "protocol" or "application"
+              +---- error-tag        string
+              +---- error-app-tag?   string
+              +---- error-path?      xpath
+              +---- error-message?   string
+
+Note: REST server will not populate error-app-tag and error-path fields in this release. It can be
+enhanced in a future release. A sample error response:
+
+    {
+      "ietf-restconf:errors" : {
+        "error" : [
+          {
+            "error-type" : "application",
+            "error-tag" : "invalid-value",
+            "error-message" : "VLAN 100 not found"
+          }
+        ]
+      }
+    }
+
+**error-type** can be either "protocol" or "application", indicating the origin of the error.
+RESTCONF defines two more error-type enums "transport" and "rpc"; they are not used by REST server.
+
+**error-tag** indicates nature of error as described in [RFC8040, section 7](https://tools.ietf.org/html/rfc8040#page-74).
+
+**error-message** field carries a human friendly error message that can be displayed to the end
+user. This is an optional field; system error do not include error-message, or have generic
+messages like "Internal error". App module developer should use human friendly messages while
+returning application errors. In case of CVL constraint violation the REST server will pick
+the error message from the YANG "error-message" statement of CVL schema YANG.
+
+Table below lists possible error conditions with response status and data returned by REST server.
+
+Method  | Error condition          | Status | error-type  | error-tag        | error-message
+--------|--------------------------|--------|-------------|------------------|----------------------
+*any*   | Incorrect request data      | 400 | protocol    | invalid-value    |
+*write* | Bad content-type            | 415 | protocol    | invalid-value    | Unsupported content-type
+*write* | OpenAPI schema validation fails | 400 | protocol| invalid-value    | Content not as per schema
```

Comment:
I'm assuming these error messages are coming from swagger - is this all the detail they provide? Can we get more context other than "the schema was wrong?" I can see this being tricky to debug for more complicated schemas.

Reply:
OpenAPI schema validation is applicable for REST APIs defined directly via an OpenAPI spec. We have very basic schema validation support here. It is possible to provide more information, but not planned for now. Payload validation for yang defined RESTCONF APIs will be done through YGOT bindings and it provides detailed error info.